### PR TITLE
Repair adzerk objects with incorrect ids

### PR DESCRIPTION
**Pull request by dwick**
Sometimes (we don't know why) adzerk will respond with an id that isn't a part
of our network.  In this case we should look up the object by name/relation and
set the id properly.  If we can't find it at all just unset it so a new object is
created.

:eyeglasses: @zeantsoi @bsimpson63

### CODE REVIEW

File: reddit_adzerk/adzerk_api.py [hunk 1]
```diff
@@ -7,16 +8,34 @@
...
+        if not isinstance(json_body, dict):
+            raise ValueError("invalid json")
+
+        if response.status_code == 400:
+            if RE_BAD_ID.match(json_body.get("message")):
+                raise AdzerkIdError()
     except ValueError:
-        raise AdzerkError('bad response')
+        if response.text and RE_BAD_ID.match(response.text):
+            raise AdzerkIdError()
```
Comment (bsimpson63) [inline reddit_adzerk/adzerk_api.py]: adzerk will give us a non-json response sometimes?

Comment (dwick) [inline reddit_adzerk/adzerk_api.py]: yes, they're returning 200s with `"... not in your network"`

json.loads("\"something\"") doesn't throw a value error, but it also just returns a unicode string.

File: reddit_adzerk/adzerkpromote.py [hunk 2]
```diff
@@ -113,6 +115,84 @@ def change_to_str(change_tuple):
     return map(change_to_str, changed)
 
 
+def _repair_bad_creative_id(link, author):
+    fixed = False
+    if not getattr(author, "external_advertiser_id", False):
+        g.log.error("no author for link %s, cannot repair" % link._fullname)
+    else:
+        all_creatives = adzerk_api.Creative.list(author.external_advertiser_id)
```
Comment (bsimpson63) [inline reddit_adzerk/adzerkpromote.py]: can you add a comment here (or somewhere else appropriate) explaining that the object exists in adzerk, but when we created it they told us the wrong id so we need to look it up.

File: reddit_adzerk/adzerkpromote.py [hunk 3]
```diff
@@ -113,6 +115,84 @@ def change_to_str(change_tuple):
...
+    if not fixed:
+        g.log.info("unsetting creative id of %d for link %s" %
+            (link.external_creative_id, link._fullname))
+        link.external_creative_id = None
+        link._commit()
+
+
+def _repair_bad_campaign_id(link):
+    fixed = False
+    campaigns_by_fullname = adzerk_api.Campaign.search(link._fullname)
```
Comment (bsimpson63) [inline reddit_adzerk/adzerkpromote.py]: `campaigns_by_fullname` sounds like it would be a dict

File: reddit_adzerk/adzerkpromote.py [hunk 4]
```diff
@@ -113,6 +115,84 @@ def change_to_str(change_tuple):
...
+    for campaign in campaigns_by_fullname:
+        if campaign.Name == link._fullname:
+            if link.external_campaign_id != campaign.Id:
```
Comment (bsimpson63) [inline reddit_adzerk/adzerkpromote.py]: why do you need this check? and why is it fixed if this is false and we don't set anything?

File: reddit_adzerk/adzerkpromote.py [hunk 5]
```diff
@@ -113,6 +115,84 @@ def change_to_str(change_tuple):
...
+    if not fixed:
+        g.log.info("unsetting flight id of %d for campaign %s" %
+            (campaign.external_flight_id, campaign._fullname))
+        campaign.external_flight_id = None
+        campaign._commit()
+
+
+def _repair_bad_advertiser_id(author):
+    fixed = False
+    advertisers_by_name = adzerk_api.Advertiser.search(author.name)
```
Comment (bsimpson63) [inline reddit_adzerk/adzerkpromote.py]: `advertisers_by_name` sounds like a dict also

**Comment by bsimpson63** [comment]
:nail_care: looks good, just a couple comments

File: reddit_adzerk/adzerkpromote.py (hunk 4 above)
Comment (dwick) [inline reddit_adzerk/adzerkpromote.py]: sometimes we get `not in your network` but the id is actually correct, can't seem to tell when though

File: reddit_adzerk/adzerk_api.py [hunk 6]
```diff
@@ -7,16 +8,34 @@
...
-        return json.loads(response.text)
+        json_body = json.loads(response.text)
+
+        if not isinstance(json_body, dict):
+            raise ValueError("invalid json")
+
+        if response.status_code == 400:
+            if RE_BAD_ID.match(json_body.get("message")):
+                raise AdzerkIdError()
     except ValueError:
```
Comment (zeantsoi) [inline reddit_adzerk/adzerk_api.py]: Is there an off-chance that `json.loads` will raise a `TypeError` if the `response.text` is empty?

Comment (dwick) [inline reddit_adzerk/adzerk_api.py]: doesn't look like it http://docs.python-requests.org/en/latest/api/#requests.Response.text

**Comment by dwick** [comment]
:haircut:

**Comment by bsimpson63** [comment]
:fish:

**Comment by dwick** [comment]
closing since we hopefully no longer need this.

**Pull request closed without merging**